### DISCUSSION TITLE
How to validate server's certificate?

### DISCUSSION THREAD

**Discussion by liwass** [microsoft/msquic#4007] [Q&A] [answered]
Hi, the client sample supplied uses unsecure configuration, that is flag QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION.
What is the proper way to make a mutual authentication, so that client validates the server's certificate?
Looks like the following 
CredConfig.Type = QUIC_CREDENTIAL_TYPE_NONE;
CredConfig.Flags = QUIC_CREDENTIAL_FLAG_CLIENT;
CredConfig.Flags |= QUIC_CREDENTIAL_FLAG_INDICATE_CERTIFICATE_RECEIVED | QUIC_CREDENTIAL_FLAG_DEFER_CERTIFICATE_VALIDATION;

triggers event QUIC_CONNECTION_EVENT_PEER_CERTIFICATE_RECEIVED, 
then 
ConnectionCertificateValidationComplete(Connection, TRUE, QUIC_TLS_ALERT_CODE_SUCCESS) 
is expected to validate the received certificate?
In any case, a segmentation fault happens in ConnectionCertificateValidationComplete (didn't dig further).

Can you shed some light on how it should be done? An updated sample would be highly appreciated.

**Comment by nibanks** [comment]
Generally, client should validate server's certificate with the default behavior automatically, so long as you don't turn off validation. You only need to mess with that other stuff if you want to do something special.

You also mention "mutual authentication" which generally means client sends a certificate too, and then the server validates it. Are you looking to do that too?

**Reply by liwass** [reply]
I see.
The issue is then when doing "secure" as in [https://github.com/microsoft/msquic/blob/main/src/tools/sample/sample.c#L761](https://github.com/microsoft/msquic/blob/main/src/tools/sample/sample.c#L761) 
there is an error:
QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_TRANSPORT, status: 0xbebc330
We are implementing a Linux SMB client over QUIC, the target server is Windows Server Datacenter Insider supporting SMB over QUIC.
Since Win11 client does fine SMB over QUIC with it, I believe the server's certificate is properly configured.

And yes, there is a scenario when we need to provide a client's certificate to the server. 
I believe this is done by using QUIC_CREDENTIAL_TYPE_CERTIFICATE_FILE and QUIC_CREDENTIAL_FLAG_SET_CA_CERTIFICATE_FILE and providing the certificate files?

**Reply by nibanks** [reply]
You're sure the server's certificate isn't self-signed and somehow the Win11 client isn't disabling server cert validation? We will likely need logs on the client to see what's going on.

**Reply by liwass** [reply]
Win11 client connects without the net use flag /skipcertcheck. 
How do we check whether it's self-signed? I can't connect with openssl s_client to this server's 443 port.

**Reply by liwass** [reply]
The server does have self-signed certificate according to IT.

**Reply by nibanks** [reply, marked as answer]
Well, by definition self-signed certs aren't valid/trusted to another machine. So you'd have to disable validation to connect.
